Userland/Libraries/LibWasm/Parser/Parser.cpp

```diff
@@ -0,0 +1,1323 @@
+
+        Vector<ResultT> entries;
+        for (size_t i = 0; i < count; ++i) {
+            auto result = T::parse(stream);
+            if (result.is_error())
+                return ParseResult<Vector<ResultT>> { result.error() };
+            entries.append(result.release_value());
+        }
+        return ParseResult<Vector<ResultT>> { move(entries) };
+    } else {
+        size_t count;
+        if (!LEB128::read_unsigned(stream, count))
+            return ParseResult<Vector<T>> { with_eof_check(stream, ParseError::ExpectedSize) };
+
+        Vector<T> entries;
+        for (size_t i = 0; i < count; ++i) {
+            if constexpr (IsSame<T, size_t>) {
+                size_t value;
+                if (!LEB128::read_unsigned(stream, value))
+                    return ParseResult<Vector<T>> { with_eof_check(stream, ParseError::ExpectedSize) };
+                entries.append(value);
+            } else if constexpr (IsSame<T, ssize_t>) {
+                ssize_t value;
+                if (!LEB128::read_signed(stream, value))
+                    return ParseResult<Vector<T>> { with_eof_check(stream, ParseError::ExpectedSize) };
+                entries.append(value);
+            } else if constexpr (IsSame<T, u8>) {
+                if (count > 64 * KiB)
+                    return ParseResult<Vector<T>> { ParseError::HugeAllocationRequested };
+                entries.resize(count);
+                if (!stream.read_or_error({ entries.data(), entries.size() }))
+                    return ParseResult<Vector<T>> { with_eof_check(stream, ParseError::InvalidInput) };
+                break; // Note: We read this all in one go!
+            }
+        }
+        return ParseResult<Vector<T>> { move(entries) };
+    }
+}
+
+static ParseResult<String> parse_name(InputStream& stream)
+{
+#if WASM_BINPARSER_DEBUG
+    ScopeLogger logger;
+#endif
+    auto data = parse_vector<u8>(stream);
+    if (data.is_error())
+        return data.error();
+
+    return String::copy(data.value());
+}
+
+template<typename T>
+struct ParseUntilAnyOfResult {
+    u8 terminator { 0 };
+    Vector<T> values;
+};
+template<typename T, typename... Args>
+static ParseResult<ParseUntilAnyOfResult<T>> parse_until_any_of(InputStream& stream, Args... terminators)
+{
+#if WASM_BINPARSER_DEBUG
+    ScopeLogger logger;
+#endif
+    ReconsumableStream new_stream { stream };
+    ScopeGuard drain_errors {
+        [&] {
+            new_stream.handle_any_error();
+        }
+    };
+
+    ParseUntilAnyOfResult<T> result;
+    for (;;) {
+        u8 byte;
+        new_stream >> byte;
+        if (new_stream.has_any_error())
+            return with_eof_check(stream, ParseError::ExpectedValueOrTerminator);
+
+        if ((... || (byte == terminators))) {
+            result.terminator = byte;
+            return result;
+        }
+
+        new_stream.unread({ &byte, 1 });
+        auto parse_result = T::parse(new_stream);
```

Comment:
Why does this function have an implicit requirement that T has a member function `T::parse(InputStream&)`, but parse_vector has a different implementation if T does not?

Reply:
Other than "there's no rule in the grammar that requires reading a bunch of sizes/bytes until a specific terminator"? None whatsoever.
Should I make that `T::parse()` requirement explicit?

Reply:
I made `parse_until_any_of<T>(InputStream&, ...)` require that T has a `parse(InputStream&)`.